src/cincinnati/client.rs

```diff
@@ -0,0 +1,124 @@
+//! Asynchronous Cincinnati client.
+
+// TODO(lucab): eventually move to its own "cincinnati client library" crate
+
+#![allow(unused)]
+
+use failure::{Error, Fallible, ResultExt};
+use futures::future;
+use futures::prelude::*;
+use reqwest::r#async as asynchro;
+use reqwest::Method;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+/// Cincinnati graph API path endpoint (v1).
+static V1_GRAPH_PATH: &str = "v1/graph";
+
+/// Cincinnati JSON protocol: node object.
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq)]
+pub struct Node {
+    pub version: String,
+    pub payload: String,
+    pub metadata: HashMap<String, String>,
+}
+
+/// Cincinnati JSON protocol: graph object.
+#[derive(Debug, Deserialize, PartialEq, Eq)]
+pub struct Graph {
+    pub nodes: Vec<Node>,
+    pub edges: Vec<(u64, u64)>,
+}
+
+/// Client to make outgoing API requests.
+#[derive(Clone, Debug)]
+pub struct Client {
+    /// Base URL for API endpoint.
+    api_base: reqwest::Url,
+    /// Asynchronous reqwest client.
+    hclient: asynchro::Client,
+    /// Client parameters (query portion).
+    query_params: HashMap<String, String>,
+}
+
+impl Client {
+    /// Fetch an update-graph from Cincinnati.
+    pub fn fetch_graph(&self) -> impl Future<Item = Graph, Error = Error> {
+        let req = self.new_request(Method::GET, V1_GRAPH_PATH);
```

Comment:
Not exactly the right place to discuss this, though I'm kinda surprised that we have to fetch the whole graph each time. Does the API not support passing the node we're currently on so it only returns a subset of the graph containing the newer node(s) that connect to it?

Reply:
The protocol is agnostic to that. We do indeed pass the `current_version` as a client parameter here. The server can trim the graph to only contain "current-version node" plus all updates (current cincinnati backend doesn't). The tradeoff here is between a larger graph to be processed by each client, and graph-filtering to be performed on the fly server-side for each request. We are currently preferring the former, as it's a bit easier to overload the server with the latter, but it's a point that we'll re-evaluate once we scale up (both in number of clients, and in releases in the graph).

Reply:
Ahh gotcha. OK that makes sense.

> We do indeed pass the `current_version` as a client parameter here.

You mean, in the actual HTTP request? Hmm, missing where that's happening here.

Reply:
Yes, as query parameter in the HTTP request. Logic is here: https://github.com/coreos/zincati/pull/15/files#diff-5f6838301c026c6c12974b331928bbb4R57.